// Copyright (c) Facebook, Inc. and its affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

syntax = "proto3";

package rbperf;

enum StorageType {
    STREAMING = 0;
    COMPACT = 1;
}

message Profile {
    enum Type {
        PERF = 0;
        TRACEPOINT = 1;
        UPROBE = 2;
        USDT = 3;
    }

    Type type = 1;
    int32 created_at = 4;
    string hostname = 5;
    string kernel_string = 6;
    string rbperf_version = 7;

    // Only used in compact mode
    repeated StackTrace stacktraces = 8;
    map<int32, string> string_table = 9;
}

message Frame {
    string path = 1;
    string method = 2;
    int32 lineno = 3;
}

message InternedFrame {
    int32 path = 1;
    int32 method = 2;
    int32 lineno = 3;
}

message StackTrace {
    enum StackStatus {
        COMPLETE = 0;
        INCOMPLETE = 1;
    }
    int64 timestamp = 1;
    int32 tid = 2;
    int32 cpu = 3;
    int32 pid = 4;
    string comm = 5;
    StackStatus stack_status = 6;
    // Ruby version?
    // Check other structs
    repeated Frame frames = 7;
    repeated InternedFrame interned_frames = 8;
}
